cpp/include/raft/matrix/detail/linewise_op.cuh

```diff
@@ -0,0 +1,516 @@
+        update = false;
+      }
+#pragma unroll VecElems
+      for (int k = 0; k < VecElems; k++, rowMod++) {
+        if (rowMod == rowLen) {
+          rowMod = 0;
+          rowDiv++;
+          int l = 0;
+          ((args[l] = vecs[rowDiv], l++), ...);
+        }
+        int l         = 0;
+        w.val.data[k] = op(v.val.data[k], (std::ignore = vecs, args[l++])...);
+      }
+      *out = w.val.internal;
+    }
+  }
+
+  /**
+   * Compute op(matrix_in, vec_1, vec_2, ...) where vectors are applied along
+   * matrix rows (vector and matrix indices are 1-1).
+   *
+   * It's assumed that `in` and `out` are aligned to the cuda-vector-size,
+   * and their length is multiple of that.
+   *
+   * Block work arrangement: striped;
+   *     the grid size is chosen in such a way, that one thread always processes
+   *     the same vector elements. That's why there is no need to read the
+   *     vector arguments multiple times.
+   *
+   * @tparam Args a pack of raft::TxN_t<Type, VecElems>
+   * @param [out] out (aligned part of) the output matrix
+   * @param [in] in (aligned part of) the input matrix
+   * @param [in] len total length of (the aligned part of) the input/output matrices
+   * @param [in] op the function to apply
+   * @param [in] args the cuda-vector-sized chunks on input vectors (raft::TxN_t<Type, VecElems>)
+   */
+  template <typename Lambda, typename... Args>
+  static __device__ __forceinline__ void vectorRows(typename Vec::io_t* out,
+                                                    const typename Vec::io_t* in,
+                                                    const IdxType len,
+                                                    Lambda op,
+                                                    Args... args) noexcept
+  {
+    Vec v;
+    const IdxType d = BlockSize * gridDim.x;
+    for (IdxType i = threadIdx.x + blockIdx.x * BlockSize; i < len; i += d) {
+      v.val.internal = __ldcv(in + i);
+#pragma unroll VecElems
+      for (int k = 0; k < VecElems; k++)
+        v.val.data[k] = op(v.val.data[k], args.val.data[k]...);
+      __stwt(out + i, v.val.internal);
+    }
+  }
+
+  /**
+   * The helper for `vectorRows`. Loads the `raft::TxN_t<Type, VecElems>` chunk
+   * of a vector. Most of the time this is not aligned, so we load it thread-striped
+   * within a block and then use the shared memory to get a contiguous chunk.
+   *
+   * @param [in] p pointer to a vector
+   * @param [in] blockOffset the offset of the current block into a vector.
+   * @param [in] rowLen the length of a vector.
+   * @return a contiguous chunk of a vector, suitable for `vectorRows`.
+   */
+  static __device__ __forceinline__ Vec loadVec(const Type* p,
+                                                const IdxType blockOffset,
+                                                const IdxType rowLen) noexcept
+  {
+    __shared__ alignas(sizeof(Type) * VecElems) Type shm[VecElems * BlockSize];
```

Comment:
It's not clear:
  * Does it allocate `shm` once, or on every function call?
  * How much memory does it use?
  * Does it synchronize the memory implicitly between the calls, if it's only one blob?

Try to move this declaration outside to the kernel and use double buffering to make sure we don't allocate too much and don't corrupt the data without `syncthreads` in between the function calls.